embeddings/task/lightning_task/sequence_labeling.py

```diff
@@ -0,0 +1,129 @@
+        return metrics
+
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+        assert (not (args and kwargs)) and (args or kwargs)
+        inputs = kwargs if kwargs else args
+        if isinstance(inputs, tuple):
+            inputs = dict(ChainMap(*inputs))
+        return self.model(**inputs)
+
+    def shared_step(self, **batch: Any) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        outputs = self.forward(**batch)
+        loss, logits = outputs[:2]
+        preds = torch.argmax(logits, dim=2)
+        return loss, logits, preds
+
+    def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
+        batch, batch_idx = args
+        labels = batch["labels"]
+        loss, logits, preds = self.shared_step(**batch)
+        self.train_metrics(preds[labels != self.ignore_index], labels[labels != self.ignore_index])
+        self.log("train/Loss", loss)
+        if self.hparams.use_scheduler:
+            assert self.trainer is not None
+            last_lr = self.trainer.lr_schedulers[0]["scheduler"].get_last_lr()
+            self.log("train/BaseLR", last_lr[0], prog_bar=True)
+            self.log("train/LambdaLR", last_lr[1], prog_bar=True)
+        return {"loss": loss}
+
+    def validation_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
+        batch, batch_idx = args
+        labels = batch["labels"]
+        loss, logits, preds = self.shared_step(**batch)
+        self.val_metrics(preds[labels != self.ignore_index], labels[labels != self.ignore_index])
+        self.log("val/Loss", loss, on_epoch=True)
+        return None
+
+    def test_step(self, *args: Any, **kwargs: Any) -> Optional[STEP_OUTPUT]:
+        batch, batch_idx = args
+        labels = batch["labels"]
+        loss, logits, preds = self.shared_step(**batch)
+        if -1 not in labels:
+            self.test_metrics(
+                preds[labels != self.IGNORE_INDEX], labels[labels != self.IGNORE_INDEX]
+            )
+            self.log("test/Loss", loss, on_epoch=True)
+        else:
+            _logger.warning("Missing labels for the test data")
+        return None
+
+    def predict(
+        self, dataloader: DataLoader[HuggingFaceDataset]
+    ) -> Dict[str, nptyping.NDArray[Any]]:
+        predictions = [self.model(**batch).logits for batch in dataloader]
+        predictions = list(torch.argmax(torch.cat(predictions), dim=2).numpy())
+        ground_truth = list(torch.cat([x["labels"] for x in dataloader]).numpy())
+
+        assert self.trainer is not None
+        for i, (pred, gt) in enumerate(zip(predictions, ground_truth)):
+            predictions[i] = [
+                self.trainer.datamodule.id_to_label[x] for x in list(pred[gt != self.ignore_index])
+            ]
+            ground_truth[i] = [
+                self.trainer.datamodule.id_to_label[x] for x in list(gt[gt != self.ignore_index])
+            ]
+
+        return {"y_pred": np.array(predictions), "y_true": np.array(ground_truth)}
```

Comment:
Are conversions in lines 117 and 118 necessary? you can iterate over torch objects

Reply:
yes, I can iterate over numpy object but I cannot do the replacement of values in numpy array (predictions[I] = ...)
nevertheless, I refactored the code little bit here